cli/flox/src/commands/services/restart.rs

```diff
@@ -21,27 +30,89 @@ pub struct Restart {
 
 impl Restart {
     #[instrument(name = "restart", skip_all)]
-    pub async fn handle(self, flox: Flox) -> Result<()> {
+    pub async fn handle(self, config: Config, flox: Flox) -> Result<()> {
         subcommand_metric!("services::restart");
 
-        let env = supported_environment(&flox, &self.environment)?;
+        let mut concrete_environment = supported_concrete_environment(&flox, &self.environment)?;
+        let activated_environments = activated_environments();
+
+        if !activated_environments.is_active(&UninitializedEnvironment::from_concrete_environment(
+            &concrete_environment,
+        )?) {
+            return Err(anyhow!(indoc! {"
+                Cannot restart services for an environment that is not activated.
+
+                To activate and start services, run 'flox activate --start-services'
+            "}));
+        }
+
+        // TODO: this doesn't need to be mut
+        let env = concrete_environment.dyn_environment_ref_mut();
         let socket = env.services_socket_path(&flox)?;
 
+        let start_new_process_compose = if !socket.exists() {
+            true
+        } else if self.names.is_empty() {
+            // TODO: We could optimise by checking whether the manifest has actually changed.
+            process_compose_down(&socket)?;
+            true
+        } else {
+            let processes = ProcessStates::read(&socket)?;
+            let all_processes_stopped = processes.iter().all(|p| p.is_stopped());
+            if all_processes_stopped {
+                process_compose_down(&socket)?;
+            }
+            all_processes_stopped
+        };
+
+        if start_new_process_compose {
+            debug!("restarting services in new process-compose instance");
+            let names = start_with_new_process_compose(
+                config,
+                flox,
+                self.environment,
+                concrete_environment,
+                &self.names,
+            )
+            .await?;
+            for name in names {
+                message::updated(format!("Service '{name}' restarted."));
```

Comment:
nit: I think here and below it might be nice to say a service was started rather than restarted if it wasn't previously running

Reply:
I'll see how much work it is a raise a follow-up PR.